### PR TITLE
aws-sdk

### PR DESCRIPTION
**Pull request by ixandidu**
Change S3Backend to use aws-sdk (ref: https://github.com/tricycle/Radiopaedia/pull/96#issuecomment-59305976)

### CODE REVIEW

File: lib/technoweenie/attachment_fu/backends/s3_backend.rb [hunk 1]
```diff
@@ -202,7 +203,17 @@ def self.included(base) #:nodoc:
           end
           base.class_eval(eval_string, __FILE__, __LINE__)
 
-          Base.establish_connection!(s3_config.slice(:access_key_id, :secret_access_key, :server, :port, :use_ssl, :persistent, :proxy))
+          # https://forums.aws.amazon.com/thread.jspa?messageID=290781#290781
```
Comment (jaischeema) [inline lib/technoweenie/attachment_fu/backends/s3_backend.rb]: Link to the reason is fine but a simple explanation in comment will make it much easier to get the information while reading the code.

File: lib/technoweenie/attachment_fu/backends/s3_backend.rb [hunk 2]
```diff
@@ -202,7 +203,17 @@ def self.included(base) #:nodoc:
...
+          AWS.eager_autoload!
+
+          AWS.config(
+            s3_config.slice(
+              :access_key_id,
+              :secret_access_key,
+              :use_ssl,
+              :proxy_url
```
Comment (jaischeema) [inline lib/technoweenie/attachment_fu/backends/s3_backend.rb]: This is actually `proxy_uri`.

**Comment by ixandidu** [comment]
@jaischeema I have pushed the required change.

**Comment by jaischeema** [comment]
Looks good :+1: but I will have to test it against the Radiopaedia app to see if it works as expected.

**Comment by wezm** [comment]
Hi Ikhsan, this is working much better now. There is still one intermittent error occurring, although it appears that when Sidekiq retries it succeeds on that that subsequent time. The Error is question is:

```
99 INFO: fail: 2.894 sec
2014-10-23T03:22:41.417Z 21416 TID-ov9m6bomg WARN: {"retry"=>8, "queue"=>"default", "backtrace"=>true, "class"=>"Sidekiq::Extensions::DelayedModel", "args"=>["---\n- !ruby/object:Image\n  attributes:\n    id: 8623678\n    original_filename: \n    created_at: &1 2014-10-23 03:22:37.294462499 Z\n    updated_at: *1\n    content_type: image/jpeg\n    description: \n    filename: 7582b47ff25563dfb3e4980c029464.jpg\n    height: 630\n    width: 630\n    stack_id: \n    position: 1\n    contributor_id: 2957\n    parent_id: \n    size: 5036\n    thumbnail: \n    aux_modality: \n    plane_projection: \n    single_or_stack_root: true\n- :generate_thumbnails\n- []\n"], "jid"=>"3104afbf237ae9b981206999", "enqueued_at"=>1414034557.3238215, "error_message"=>"No Such Key", "error_class"=>"AWS::S3::Errors::NoSuchKey", "failed_at"=>1414034561.3436258, "retry_count"=>0, "error_backtrace"=>["/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/core/client.rb:375:in `return_or_raise'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/core/client.rb:476:in `client_request'", "(eval):3:in `get_object'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/s3/s3_object.rb:1372:in `get_object'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/s3/s3_object.rb:1091:in `read'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/bundler/gems/attachment_fu-79e482d675cb/lib/technoweenie/attachment_fu/backends/s3_backend.rb:333:in `create_temp_file'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/bundler/gems/attachment_fu-79e482d675cb/lib/technoweenie/attachment_fu.rb:318:in `generate_thumbnails'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/extensions/active_record.rb:19:in `perform'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:52:in `block (2 levels) in process'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:122:in `call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:122:in `block in invoke'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/active_record.rb:6:in `call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/retry_jobs.rb:74:in `call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/logging.rb:11:in `block in call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/logging.rb:22:in `with_context'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/logging.rb:7:in `call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:127:in `call'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:127:in `invoke'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:51:in `block in process'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:94:in `stats'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:50:in `process'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `public_send'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `dispatch'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:122:in `dispatch'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:322:in `block in handle_message'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:416:in `block in task'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:55:in `block in initialize'", "/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks/task_fiber.rb:13:in `block in create'"]}
2014-10-23T03:22:43.278Z 21416 TID-ov9m6bomg WARN: No Such Key
2014-10-23T03:22:43.278Z 21416 TID-ov9m6bomg WARN: /srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/core/client.rb:375:in `return_or_raise'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/core/client.rb:476:in `client_request'
(eval):3:in `get_object'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/s3/s3_object.rb:1372:in `get_object'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/aws-sdk-v1-1.55.0/lib/aws/s3/s3_object.rb:1091:in `read'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/bundler/gems/attachment_fu-79e482d675cb/lib/technoweenie/attachment_fu/backends/s3_backend.rb:333:in `create_temp_file'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/bundler/gems/attachment_fu-79e482d675cb/lib/technoweenie/attachment_fu.rb:318:in `generate_thumbnails'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/extensions/active_record.rb:19:in `perform'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:52:in `block (2 levels) in process'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:122:in `call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:122:in `block in invoke'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/active_record.rb:6:in `call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/retry_jobs.rb:74:in `call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/logging.rb:11:in `block in call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/logging.rb:22:in `with_context'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/server/logging.rb:7:in `call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:127:in `call'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/middleware/chain.rb:127:in `invoke'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:51:in `block in process'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:94:in `stats'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/processor.rb:50:in `process'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `public_send'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `dispatch'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:122:in `dispatch'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:322:in `block in handle_message'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:416:in `block in task'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:55:in `block in initialize'
/srv/www/radiopaedia-staging/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks/task_fiber.rb:13:in `block in create'
2014-10-23T03:22:43.527Z 21416 TID-ov9m6bomg Sidekiq::Extensions::DelayedModel JID-9f36a1d45603a6b676af0c0e INFO: start
```

**Comment by ixandidu** [comment]
I think it's happened because the worker has started doing its job while the file has not completely uploaded to s3. We might need to change [this line](https://github.com/tricycle/attachment_fu/blob/aws-sdk/lib/technoweenie/attachment_fu.rb#L504) to use `delay_for` to avoid the Error and probably add optional key to configure the interval in the `attachment_options`.

**Comment by wezm** [comment]
Yes that would make sense. I’m don’t like adding arbitrary delays though. It should be possible to queue the processing job after the upload finishes right? I wonder why this wasn’t an issue with delayed_job.

**Comment by ixandidu** [comment]
> It should be possible to queue the processing job after the upload finishes right? 

Yes, I think it is possible by moving the `save_to_storage` call before the `generate_thumbnails` call (https://github.com/tricycle/attachment_fu/blob/aws-sdk/lib/technoweenie/attachment_fu.rb#L503-L530)

> I wonder why this wasn’t an issue with delayed_job.

Probably because DJ sleep for 5 second: https://github.com/collectiveidea/delayed_job/blob/62c929abfbb1548822ff9c31d806d2ade1bec528/lib/delayed/worker.rb#L12

**Comment by jaischeema** [comment]
Merged into aws-sdk branch, thanks for that. I will test this and see how we go.

**Pull request closed without merging**